llvm-project/Dockerfile.epoch4

```diff
@@ -0,0 +1,300 @@
+  done
+
+### Linux
+COPY --from=source linux-5.18-rc6.tar.gz .
+RUN tar xf linux-5.18-rc6.tar.gz
+RUN apk add make musl-dev rsync
+RUN make -C linux-5.18-rc6 INSTALL_HDR_PATH=/sysroot/usr LLVM=1 -j$(nproc) headers_install
+RUN apk del rsync musl-dev make
+
+### Musl
+COPY --from=source musl-1.2.3.tar.gz .
+RUN tar xf musl-1.2.3.tar.gz
+ARG MUSL_DIR=musl-1.2.3/build
+RUN mkdir -p ${MUSL_DIR}
+RUN cd ${MUSL_DIR} && \
+  CC=clang AR=llvm-ar RANLIB=llvm-ranlib \
+  ../configure --prefix=/usr --syslibdir=/usr/lib
+RUN apk add make
+RUN make -C ${MUSL_DIR} -j$(nproc)
+RUN make -C ${MUSL_DIR} -j$(nproc) DESTDIR=/sysroot install-headers
+RUN make -C ${MUSL_DIR} -j$(nproc) DESTDIR=/sysroot install-libs
+RUN apk del make
+
+# Pause for a quick sanity check
+COPY hello.c hello.cpp /
+ARG SYSROOT=--sysroot=/sysroot
+RUN clang ${SYSROOT} hello.c && ./a.out && \
+  clang ${SYSROOT} hello.c -static && ./a.out && \
+  clang++ ${SYSROOT} hello.cpp && ./a.out && \
+  clang++ ${SYSROOT} hello.cpp -static -lc++abi && ./a.out
+
+### Zlib
+COPY --from=source zlib-1.2.12.tar.gz .
+RUN tar xf zlib-1.2.12.tar.gz
+ARG ZLIB_DIR=zlib-1.2.12/build
+RUN mkdir -p ${ZLIB_DIR}
+RUN cd ${ZLIB_DIR} && \
+  CC="clang ${SYSROOT}" AR=llvm-ar ../configure --prefix=/sysroot/usr
+RUN apk add make
+RUN make -C ${ZLIB_DIR} -j$(nproc)
+RUN make -C ${ZLIB_DIR} -j$(nproc) install
+RUN apk del make
+
+### LLVM
+COPY --from=source llvm-project-14.0.1.src.tar.xz .
+RUN tar xf llvm-project-14.0.1.src.tar.xz && \
+  mv llvm-project-14.0.1.src llvm-project
+RUN apk add cmake ninja python3
+COPY stage4.cmake llvm-project/.
+ARG LLVM_BUILD_DIR=llvm-project/llvm/build
+RUN cmake \
+  -B ${LLVM_BUILD_DIR} \
+  -C llvm-project/stage4.cmake \
+  -D LLVM_DEFAULT_TARGET_TRIPLE=$(clang -print-target-triple) \
+  -S llvm-project/llvm \
+  -G Ninja
+RUN ninja -C ${LLVM_BUILD_DIR} install-clang install-lld
+RUN ninja -C ${LLVM_BUILD_DIR} install-clang-resource-headers
+RUN ninja -C ${LLVM_BUILD_DIR} \
+  install-llvm-ar \
+  install-llvm-nm \
+  install-llvm-objcopy \
+  install-llvm-objdump \
+  install-llvm-ranlib \
+  install-llvm-readelf \
+  install-llvm-strip
+RUN ninja -C ${LLVM_BUILD_DIR} install-compiler-rt
+RUN ninja -C ${LLVM_BUILD_DIR} install-llvm-profdata
+
+RUN apk del cmake ninja python3
+
+# Final test
+RUN llvm-readelf -p .comment $(which clang) | grep -e clang -e LLD
+RUN llvm-readelf -p .comment $(which clang) | grep -v GCC
+
+# Throw away profiles from building LLVM itself.
+RUN find ${LLVM_BUILD_DIR} -name \*.profraw | xargs rm
+
+### Start doing kernel builds
+# TODO: objtool needs libelf.h and gelf.h? elfutils-dev is the alpine package.
+RUN apk add make flex bison
+ARG arch=arm
+RUN ARCH=${arch} HOSTCFLAGS=${SYSROOT} HOSTLDFLAGS=${SYSROOT} \
+      make -C linux-5.18-rc6 LLVM=1 -s -j$(nproc) allnoconfig all && \
+    llvm-profdata merge --output=${arch}.profdata.prof \
+      $(find ${LLVM_BUILD_DIR} -name \*.profraw) && \
+    find ${LLVM_BUILD_DIR} -name \*.profraw | xargs rm
```

Comment:
What's the reasoning behind merging the profiles after each kernel build then merging them all at the end? Wouldn't be be simpler to merge them all at the end?

Additionally, if these remain around, I think you can avoid the `find` call, as the `.profraw` files should be in `${LLVM_BUILD_DIR}/profiles`.

```
rm ${LLVM_BUILD_DIR}/profiles/*.profraw
```